src/visidentifier/pocowsclient.cpp

```diff
@@ -0,0 +1,261 @@
+
+    LOG_DBG() << "Got server response: requestId = " << requestId.c_str() << ", response = " << response.c_str();
+
+    return {response.cbegin(), response.cend()};
+}
+
+void PocoWSClient::SendMessage(const ByteArray& message)
+{
+    if (message.empty()) {
+        return;
+    }
+
+    std::lock_guard lock(mMutex);
+
+    if (!mIsConnected) {
+        throw WSException("Not connected", AOS_ERROR_WRAP(aos::ErrorEnum::eFailed));
+    }
+
+    try {
+        mWebSocket->setSendTimeout(Poco::Timespan(mWSConfig.mWebSocketTimeoutMiliseconds));
+
+        const int len = mWebSocket->sendFrame(&message.front(), message.size(), Poco::Net::WebSocket::FRAME_TEXT);
+
+        LOG_DBG() << "Sent " << len << "/" << message.size() << " bytes.";
+    } catch (const std::exception& e) {
+        mWSClientErrorEvent.Set(WSClientEvent::EventEnum::FAILED, e.what());
+
+        throw WSException(e.what(), AOS_ERROR_WRAP(aos::ErrorEnum::eFailed));
+    }
+}
+
+PocoWSClient::~PocoWSClient()
+{
+    Close();
+    StopReceiveFramesThread();
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+void PocoWSClient::HandleResponse(const std::string& frame)
+{
+    try {
+        const auto object = UtilsJson::ParseJson(frame).extract<Poco::JSON::Object::Ptr>();
+
+        if (object.isNull()) {
+            return;
+        }
+
+        if (const auto action = object->get(cActionTagName); action == "subscription") {
+            mHandleSubscription(frame);
+
+            return;
+        }
+
+        const auto requestId = object->get(cRequestIdTagName).convert<std::string>();
+        if (requestId.empty()) {
+            throw AosException("invalid requestId tag received");
+        }
+
+        if (!mPendingRequests.SetResponse(requestId, frame)) {
+            mHandleSubscription(frame);
+        }
+    } catch (const Poco::Exception& e) {
+        LOG_ERR() << AosException(e.what(), AOS_ERROR_WRAP(aos::ErrorEnum::eFailed)).what();
+    }
+}
+
+void PocoWSClient::ReceiveFrames()
+{
+    LOG_DBG() << "PocoWSClient::ReceiveFrames has been started.";
+
+    try {
+        int  flags;
+        int  n;
+        char buffer[1024];
+
+        do {
+            n = mWebSocket->receiveFrame(buffer, sizeof(buffer), flags);
+
+            LOG_DBG() << "PocoWSClient::ReceiveFrames recived " << n << " bytes. Flags = " << flags;
+
+            if ((flags & Poco::Net::WebSocket::FRAME_OP_BITMASK) == Poco::Net::WebSocket::FRAME_OP_CLOSE) {
+                mWSClientErrorEvent.Set(WSClientEvent::EventEnum::FAILED, "got Close frame from server");
+
+                return;
+            }
+
+            if (n > 0) {
+                const std::string message(buffer, n);
+
+                HandleResponse(message);
+            }
```

Comment:
I'm a bit confused - is 1024 the maximum frame size? Where did you get that size from? And what will happen if a JSON larger than 1024 is sent? In our case, will that be an error?

Reply:
Refactored. Used `int receiveFrame(Poco::Buffer<char>& buffer, int& flags);` overload
It's much safer

`		/// Receives a frame from the socket and stores it
		/// after any previous content in buffer.
		/// The buffer will be grown as necessary.
		///
		/// See the note below if you want the received data
		/// to be stored at the beginning of the buffer.
		///
		/// The frame's payload size must not exceed the
		/// maximum payload size set with setMaxPayloadSize().
		/// If it does, a WebSocketException (WS_ERR_PAYLOAD_TOO_BIG)
		/// is thrown and the WebSocket connection must be
		/// terminated.
		///
		/// A WebSocketException will also be thrown if a malformed
		/// or incomplete frame is received.
		///
		/// If this method is used, a reasonable maximum payload size should
		/// be set with setMaxPayloadSize() to prevent a potential
		/// DoS attack (memory exhaustion) by sending a WebSocket frame
		/// header with a huge payload size.
		///
		/// Returns the number of payload bytes received.
		/// A return value of 0, with flags also 0, means that the peer has
		/// shut down or closed the connection.
		/// A return value of 0, with non-zero flags, indicates an
		/// reception of an empty frame (e.g., in case of a PING).
		///
		/// Throws a TimeoutException if a receive timeout has
		/// been set and nothing is received within that interval.
		/// Throws a NetException (or a subclass) in case of other errors.
		///
		/// The frame flags and opcode (FrameFlags and FrameOpcodes)
		/// is stored in flags.
		///
		/// Note: Since the data received from the WebSocket is appended
		/// to the given Poco::Buffer, the buffer passed to this method
		/// should be created with a size of 0, or resize(0) should be
		/// called on the buffer beforehand, if the expectation is that
		/// the received data is stored starting at the beginning of the
		/// buffer.`